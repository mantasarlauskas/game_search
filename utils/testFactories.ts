import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
