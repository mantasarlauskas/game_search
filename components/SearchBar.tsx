import { ChangeEvent, useEffect, useRef, useState, KeyboardEvent } from 'react';
import debounce from 'debounce';
import Link from 'next/link';
import { useRouter } from 'next/router';
import styles from 'styles/search-bar.module.scss';
import { API_PATH, fetchData } from 'utils/fetch';
import { roundNumber } from 'utils/number';
import Spinner from 'components/Spinner';
import SearchIcon from 'assets/search.svg';
import { SearchResult } from 'utils/types';
import { Route } from 'utils/routes';

function SearchBar() {
    const router = useRouter();
    const ref = useRef<HTMLInputElement>(null);
    const [isOpen, setIsOpen] = useState(false);
    const [loading, setLoading] = useState(false);
    const [results, setResults] = useState<SearchResult[]>([]);

    useEffect(() => {
        function handleClick({ target }: MouseEvent) {
            if (!ref.current?.contains(target as HTMLElement)) {
                setIsOpen(false);
                ref.current?.blur();
            }
        }

        window.addEventListener('click', handleClick);
        return () => {
            window.removeEventListener('click', handleClick);
        };
    }, []);

    const handleChange = debounce(async ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
        setLoading(true);
        const data = await fetchData(API_PATH.GAMES, {
            search: value,
            page: 1,
            page_size: 5,
        });

        setResults(data.results);
        setLoading(false);
    }, 200);

    function handleKeyDown(e: KeyboardEvent<HTMLInputElement>) {
        if (e.key === 'Enter') {
            const target = e.target as HTMLInputElement;
            router.push(`/search/${target.value}`);
            setIsOpen(false);
            ref.current?.blur();
        }
    }

    return (
        <div className={styles.root}>
            <div className={styles.inputWrapper}>
                <input
                    ref={ref}
                    className={styles.input}
                    onChange={handleChange}
                    onKeyDown={handleKeyDown}
                    onFocus={() => setIsOpen(true)}
                    placeholder="Search games"
                />
                <SearchIcon className={styles.icon} />
            </div>
            {(loading || results.length > 0) && isOpen && (
                <div className={styles.results}>
                    {loading ? <Spinner /> : results.map(({
                        name,
                        slug,
                        background_image,
                        rating,
                    }) => (
                        <Link key={slug} href={`${Route.GAMES}/${slug}`}>
                            <div className={styles.result}>
                                <div
                                    className={styles.image}
                                    style={{ backgroundImage: `url(${background_image})` }}
                                />
                                <div>
                                    <div className={styles.name}>{name}</div>
                                    <div className={styles.rating}>{roundNumber(rating)}</div>
                                </div>
                            </div>
                        </Link>
                    ))}
                </div>
            )}
        </div>
    );
}

export default SearchBar;
