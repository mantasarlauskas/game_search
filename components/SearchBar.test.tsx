import { fireEvent, render, screen } from '@testing-library/react';
import SearchBar from 'components/SearchBar';
import { getGames } from 'testing/testFactories';
import userEvent from '@testing-library/user-event';
import { useQuery } from 'react-query';

jest.mock('components/Spinner', () => () => <span>Spinner</span>);
jest.mock('react-query', () => ({
    useQuery: jest.fn(),
}));

const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter() {
        return {
            push: mockPush,
        };
    },
}));

(useQuery as jest.Mock).mockReturnValue({
    data: { results: getGames() },
    isLoading: false,
});

describe('<SearchBar />', () => {
    it('matches snapshot', () => {
        const { container } = render(<SearchBar />);
        expect(container).toMatchSnapshot();
    });

    it('updates input value on change', () => {
        render(<SearchBar />);
        userEvent.type(screen.getByPlaceholderText('Search games'), 'text');
        expect((screen.getByRole('textbox') as HTMLInputElement).value).toEqual('text');
    });

    it('shows games on input click and hides them on outside click', () => {
        render(<SearchBar />);
        expect(screen.queryByText('Portal 2')).not.toBeInTheDocument();
        expect(screen.queryByText('Tomb Raider (2013)')).not.toBeInTheDocument();

        userEvent.click(screen.getByRole('textbox'));
        expect(screen.getByText('Portal 2')).toBeInTheDocument();
        expect(screen.getByText('Tomb Raider (2013)')).toBeInTheDocument();

        userEvent.click(document.body);
        expect(screen.queryByText('Portal 2')).not.toBeInTheDocument();
        expect(screen.queryByText('Tomb Raider (2013)')).not.toBeInTheDocument();
    });

    it('on game click goes to game page', () => {
        render(<SearchBar />);
        userEvent.click(screen.getByRole('textbox'));
        userEvent.click(screen.getByText('Portal 2'));
        expect(window.location.href).toEqual('/games/portal-2');
    });

    it('on enter click goes to search page', () => {
        render(<SearchBar />);
        userEvent.type(screen.getByPlaceholderText('Search games'), 'text');
        fireEvent.keyDown(screen.getByRole('textbox'), { key: 'Enter' });
        expect(mockPush).toBeCalledTimes(1);
        expect(mockPush).toBeCalledWith('/search/text');
    });

    it('shows spinner if data is fetching', () => {
        (useQuery as jest.Mock).mockReturnValue({ isLoading: true });
        render(<SearchBar />);
        userEvent.click(screen.getByRole('textbox'));
        expect(screen.getByText('Spinner')).toBeInTheDocument();
    });
});
