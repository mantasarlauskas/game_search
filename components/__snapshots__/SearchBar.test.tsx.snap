// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="inputWrapper"
    >
      <input
        class="input"
        placeholder="Search games"
        value=""
      />
      <span />
    </div>
  </div>
</div>
`;
